### ISSUE TITLE
Feature request: compile-time resolution of "super-local" label

### ISSUE THREAD

**Issue by mkeeter** [CensoredUsername/dynasm-rs#69]
I noticed that hashmap lookups are taking a decent amount of JIT time when using local labels.

For example, if I manually compute jumps in this code:
```rust
dynasm!(ops
    // Basically the same as MinRegReg
    ; zip2 v4.s2, V(lhs_reg).s2, V(rhs_reg).s2
    ; zip1 v5.s2, V(rhs_reg).s2, V(lhs_reg).s2
    ; fcmgt v5.s2, v5.s2, v4.s2
    ; fmov x15, d5

    ; tst x15, #0x1_0000_0000
    ; b.ne >lhs

    ; tst x15, #0x1
    ; b.eq >both

    // LHS < RHS
    ; fmov D(out_reg), D(rhs_reg)
    ; mov w16, #CHOICE_RIGHT
    ; b >end

    // RHS < LHS
    ;lhs:
    ; fmov D(out_reg), D(lhs_reg)
    ; mov w16, #CHOICE_LEFT
    ; b >end

    ;both:
    ; fmax V(out_reg).s2, V(lhs_reg).s2, V(rhs_reg).s2
    ; mov w16, #CHOICE_BOTH

    ;end:
    ; strb w16, [x0], #1 // post-increment
)
```
I end up with something like this:
```rust
dynasm!(ops
    ; zip2 v4.s2, V(lhs_reg).s2, V(rhs_reg).s2
    ; zip1 v5.s2, V(rhs_reg).s2, V(lhs_reg).s2
    ; fcmgt v5.s2, v5.s2, v4.s2
    ; fmov x15, d5

    ; tst x15, #0x1_0000_0000
    ; b.ne #24 // -> lhs

    ; tst x15, #0x1
    ; b.eq #28 // -> both

    // LHS < RHS
    ; fmov D(out_reg), D(rhs_reg)
    ; mov w16, #CHOICE_RIGHT
    ; b #24 // -> end

    // <- lhs (when RHS < LHS)
    ; fmov D(out_reg), D(lhs_reg)
    ; mov w16, #CHOICE_LEFT
    ; b #12 // -> end

    // <- both
    ; fmax V(out_reg).s2, V(lhs_reg).s2, V(rhs_reg).s2
    ; mov w16, #CHOICE_BOTH

    // <- end
    ; strb w16, [x0], #1 // post-increment
)
```

In my codebase, this reduces the time spent in `dynasm` by about 30%, which is a decent chunk of performance!

It would be great to introduce a new flavor of label which is only valid during a **single** `dynasm!` block; the branch offset could then be computed at compile-time instead of runtime.

**Comment by Techcable** [comment]
Yes. Keep in mind that the original DynASM project for C (used as backend in LuaJIT 1.x, used as frontend in ) does *not* use a hash map for labels. It is a plain array.

I know this is hard given the constraints of rust proc-macros, but Ideally we would move in that direction......

**Comment by CensoredUsername** [comment]
It wouldn't necessarily have to be a new type of label, it'd be possible to guarantee that local labels in a single block always get determined at compile time. It'll be quite annoying to implement though.

But before we have a try at that, The default `LabelRegistry` just uses the standard cryptographically secure hasher in the label hashMaps. You could try benchmarking it with `FnvHasher` instead.

**Comment by CensoredUsername** [comment]
Idea: it wouldn't be hard to just special case strings of length 1 to just be array lookups instead. Make those bypass the internal hashmaps, if recent changes do not alleviate the bottleneck enough.

**Comment by CensoredUsername** [comment]
As discussed in the related pull request, you can use dynamic labels to skip the overhead from hashmap lookups if you want to know the theoretical speedup due to that. That would be useful knowledge to have before proceeding on working on this.

**Comment by CensoredUsername** [comment]
Closing this due to inactivity after requested information.

**Comment by CensoredUsername** [comment]
I've thought a bit more about this.

The important thing to push these costs to compile time would be for there to be absolutely no variable-width items inbetween. So no directives, no code that could push anything further.

Then that would allow these labels to be resolved to simple immediate expressions at compile time. This would have to be resolved at serialisation time as only at that point do we know the exact length of the instruction stream.

That said, that is a lot of work for a fairly minor benefit. You say there's a large cost but really per label we're talking like 10-15 instructions saved maybe.  Before working on that I'd really want to see some benchmarks that show that it's actually worth it. Like one comparing standard local labels vs dynamic labels vs just inserting an immediate into the jump.
